internal/clients/kibana/alerting.go

```diff
@@ -0,0 +1,181 @@
+package kibana
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/elastic/terraform-provider-elasticstack/generated/alerting"
+	"github.com/elastic/terraform-provider-elasticstack/internal/clients"
+	"github.com/elastic/terraform-provider-elasticstack/internal/models"
+	"github.com/elastic/terraform-provider-elasticstack/internal/utils"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+)
+
+func ruleResponseToModel(spaceID string, res *alerting.RuleResponseProperties) *models.AlertingRule {
+	if res == nil {
+		return nil
+	}
+
+	actions := []models.AlertingRuleAction{}
+	for _, action := range res.Actions {
+		actions = append(actions, models.AlertingRuleAction{
+			Group:  *action.Group,
+			ID:     *action.Id,
+			Params: action.Params,
+		})
+	}
+
+	scheduledTaskId := ""
+	if res.ScheduledTaskId != nil {
+		scheduledTaskId = *res.ScheduledTaskId
+	}
+
+	return &models.AlertingRule{
+		ID:         res.Id,
+		SpaceID:    spaceID,
+		Name:       res.Name,
+		Consumer:   res.Consumer,
+		NotifyWhen: string(*res.NotifyWhen),
+		Params:     res.Params,
+		RuleTypeID: res.RuleTypeId,
+		Schedule: models.AlertingRuleSchedule{
+			Interval: *res.Schedule.Interval,
+		},
+		Enabled:         &res.Enabled,
+		Tags:            res.Tags,
+		Throttle:        res.Throttle.Get(),
+		ScheduledTaskID: scheduledTaskId,
+		ExecutionStatus: models.AlertingRuleExecutionStatus{
+			LastExecutionDate: res.ExecutionStatus.LastExecutionDate.String(),
+			Status:            *res.ExecutionStatus.Status,
+		},
+		Actions: actions,
+	}
+}
+
+func PutAlertingRule(ctx context.Context, apiClient *clients.ApiClient, isCreate bool, rule models.AlertingRule) (*models.AlertingRule, diag.Diagnostics) {
+	client, err := apiClient.GetAlertingClient()
+	if err != nil {
+		return nil, diag.FromErr(err)
+	}
+
+	ctxWithAuth := apiClient.SetAlertingAuthContext(ctx)
+
+	actions := []alerting.ActionsInner{}
+	for _, action := range rule.Actions {
+		actions = append(actions, alerting.ActionsInner{
+			Group:  &action.Group,
+			Id:     &action.ID,
+			Params: action.Params,
+		})
+	}
+
+	if isCreate {
+		reqModel := alerting.CreateRuleRequest{
+			Consumer:   rule.Consumer,
+			Actions:    actions,
+			Enabled:    rule.Enabled,
+			Name:       rule.Name,
+			NotifyWhen: (*alerting.NotifyWhen)(&rule.NotifyWhen),
+			Params:     rule.Params,
+			RuleTypeId: rule.RuleTypeID,
+			Schedule: alerting.Schedule{
+				Interval: &rule.Schedule.Interval,
+			},
+			Tags:     rule.Tags,
+			Throttle: *alerting.NewNullableString(rule.Throttle),
+		}
+		req := client.CreateRule(ctxWithAuth, rule.SpaceID, "").KbnXsrf("true").CreateRuleRequest(reqModel)
+		ruleRes, res, err := req.Execute()
+		if err != nil && res == nil {
+			return nil, diag.FromErr(err)
+		}
+
+		defer res.Body.Close()
+		return ruleResponseToModel(rule.SpaceID, ruleRes), utils.CheckHttpError(res, "Unabled to create alerting rule")
```

Comment:
It looks like we don't use the result model besides its id. The same goes for Update. Does it makes sense to return only id and error?

Reply:
Yea, I thought about doing this. I don't have a strong justification either way, I decided to return the model since it was consistent with the API behaviour, rather than just returning what the current consumer wanted 🤷 happy to change it if you feel strongly.